ens-improvement-proposals/ensip-15-normalization-standard.md

```diff
@@ -0,0 +1,467 @@
+	* The 3rd and 4th characters must not both be `2D (-) HYPHEN-MINUS`.
+		* Must not match `/^..--/`
+		* Examples: `"ab-c"` and `"---a"`are valid, `"xn--"` and `----` are invalid.
+	* Return `"ASCII"`
+		* The label is free of **Fenced** and **Combining Mark** characters, and not confusable.
+1. Concatenate all the tokens together.
+	* `5F (_) LOW LINE` can only occur at the start.
+	* The first and last characters cannot be **Fenced**.
+		* Examples: `"a’s"` and `"a・a"` are valid, `"’85"` and `"joneses’"` and `"・a・"` are invalid.
+	* **Fenced** characters cannot be contiguous.
+		* Examples: `"a・a’s"` is valid, `"6’0’’"` and `"a・・a"` are invalid.
+1. The first character of every `Text` token must not be a **Combining Mark**.
+1. Concatenate the `Text` tokens together.
+1. Find the first **Group** that contain every text character:
+	* If no group is found, the label cannot be normalized.
+1. If the group is not **CM Whitelisted**:
+	* Apply NFD to the concatenated text characters.
+	* For every contiguous sequence of **NSM** characters:
+		* Each character must be unique.
+			* Example: `"x̀̀" [78 300 300]` has (2) grave accents.
+		* The number of **NSM** characters cannot exceed **Maximum NSM** (4).
+			* Example: ` "إؐؑؒؓؔ"‎ [625 610 611 612 613 614]` has (6) **NSM**.
+1. [Wholes](#wholes) — check if text characters form a confusable.
+1. The label is valid.
+	* Return the name of the group as the **Label Type**.
+
+Examples:
+
+1. `Emoji["💩️"]` + `Emoji["💩️"]` → `"Emoji"`
+1. `Text["abc$123"]` → `"ASCII"`
+1. `Emoji["🚀️"]` + `Text["à"]` → `"Latin"`
+
+### Wholes
+
+A label is [whole-script confusable](https://unicode.org/reports/tr39/#def_whole_script_confusables) if a similarly-looking valid label can be constructed using one alternative character from a different group.  The complete list of **Whole Confusables** can be found in [`spec.json`](#description-of-specjson).  Each **Whole Confusable** has a set of non-confusing characters (`"valid"`) and a set of confusing characters (`"confused"`) where each character may be the member of one or more groups.
+
+Example: **Whole Confusable** for `"g"`
+
+| Type | Code | Form | Character  | Latn | Hani | Japn | Kore | Armn | Cher | Lisu |
+| :-: | -: | :-: | :- | :-: | :-: | :-: | :-: | :-: | :-: | :-: |
+| valid | `67` | `g` | LATIN SMALL LETTER G | A | A | A | A |
+| confused | `581` | `ց` | ARMENIAN SMALL LETTER CO  | | | | | B | 
+| confused | `13C0` | `Ꮐ` | CHEROKEE LETTER NAH  | | | | | | C | 
+| confused | `13F3` | `Ᏻ` | CHEROKEE LETTER YU  |	| | | | | C |
+| confused |  `A4D6` | `ꓖ` | LISU LETTER GA | | | | | | | D |
+
+1. Allocate an empty character buffer.
+1. Start with the set of **ALL** groups.
+1. For each unique character in the label:
+	* If the character is **Confused** (a member of a **Whole Confusable**):
+		* Retain groups with **Whole Confusable** characters excluding the **Confusable Extent** of the matching **Confused** character.

```

Comment:
As an implementer, I'm having a difficult time walking through this section. I'm not sure what this line means exactly. I think giving one example flow might help or maybe a bit better clarification.

Reply:
This algorithm was difficult to explain, but it's very easy to understand once you work through a few examples or see it visually.   This tool may help: https://adraffy.github.io/ens-normalize.js/test/confused.html

The Unicode suggestion is also [very terse](https://www.unicode.org/reports/tr39/#Whole_Script_Confusables) and suffers from the problem that it's only hypothetical: *"As usual, this algorithm is intended only as a definition; implementations should use an optimized routine that produces the same result."*  It also doesn't determine how to choose a "winner" and interacts weirdly with the Augmented Scripts and Script Extensions.   From my POV, this is why this logic is implemented virtually no where on the web.

For reference, this is [my implementation](https://github.com/adraffy/ens-normalize.js/blob/main/src/lib.js#L353), which is only a few lines.   My code uses an additional optimization which is the "Confusable Extent Complement", so instead of retaining the excluded, you just intersect a pre-computed complement (for every confusable character, I've precomputed the groups that are outside that characters confusable extent.)

---

Another way of describing the algorithm is that you're trying to create a function which generates confusable strings.  You already know the string you have is "valid" (according to the previous step of Validate: every character belongs to at least one group.)  

Starting from a generator of all strings, you take each character in the actual string, and fine-tune the generator to produce similar strings.  However, instead of making it only generate the original string, you make it generate all of the OTHER (confusable) strings EXCEPT original string.  

If you can successfully produce a function of this kind, then the original string is confusable by construction — there exists a confusable string that isn't linked to the original string (or its extents.)

Since some groups have confusables between their own characters and some characters are shared between groups, the Confusable Extent concept is necessary to ensure that the constructed generator doesn't produce strings that are just valid-but-different substitutions of the original string.

---

> Retain groups with **Whole Confusable** characters excluding the **Confusable Extent** of the matching **Confused** character.

> "Retain a set of the groups from all Confusable Extents excluding the groups in the Confusable Extent of the matching character." Is that what this is asking for?

The extent is only relevant on the current confusable to exclude everything it could be substituted with it.

Prior Groups: `{A, B, C, D}`
Current Character: `X` (placeholder)
Current Confusable: `X → {C, D}, Y → {B}` 
Union of Confusable Groups: `{B, C, D}`
Current Character Confusable Extent: `{X} x {C, D}`
Confusable Groups not in Extent: `{B}`
Retained Groups: `{B}`